### PR TITLE
Move API branch into main branch + synth gen from data

### PR DESCRIPTION
**Pull request by JGSweets**
This PR does two things:
1. takes the refactor api and moves it into the main API.
2. Allows synthetic data to be generated from data w/o a report

### CODE REVIEW

File: Synthetic_Data_Generators.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+A data profile is a statistical overview of a raw data file. Each profile describes a particular type of raw data (tabular, graph, or unstructured) and therefore must be treated differently in the synthetic data generation process. This API is designed to differentially synthesize data in object-oriented fashion (i.e., with classes) based on the type of data described by its input profile.
```
Comment (JGSweets) [inline Synthetic_Data_Generators.md:1]: Refactored, but from redesign branch.

File: synthetic_data/base_generator.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+"""Contains abstract class from which generators will inherit."""
```
Comment (JGSweets) [inline synthetic_data/base_generator.py:1]: Copy from redesign branch.

File: synthetic_data/generator_builder.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+"""Contains generator factory class that picks generator based on its profile type."""
```
Comment (JGSweets) [inline synthetic_data/generator_builder.py:1]: Mostly a copy from redesign api branch

File: synthetic_data/generator_builder.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    def __new__(cls, seed=None, config=None, *args, **kwargs):
+        """Instantiate proper generator."""
+        if config:
+            try:
+                return config(*args, **kwargs)
+            except Exception as e:
+                raise ValueError("Warning: profile doesn't match user setting.")
+
+        profile = kwargs.pop("profile", None)
+        data = kwargs.pop("data", None)
```
Comment (JGSweets) [inline synthetic_data/generator_builder.py:34]: added this data kwarg

File: synthetic_data/base_generator.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+class BaseGenerator(metaclass=abc.ABCMeta):
+    """Abstract generator class."""
+
+    def __init__(self, profile, seed=None, *args, **kwargs):
+        """Initialize generator object."""
+        self.seed = seed
+        self.profile = profile
+
+    @classmethod
+    def post_profile_processing_w_data(cls, data, profile):
```
Comment (JGSweets) [inline synthetic_data/base_generator.py:15]: added new

File: synthetic_data/generator_builder.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        if not profile and not data:
+            raise ValueError(
+                "No profile object or dataset was passed in kwargs. "
+                "If you want to generate synthetic data from a "
+                "profile, pass in a profile object through the "
+                'key "profile" or data through the key "data" in kwargs.'
+            )
+
+        if data is not None:
```
Comment (JGSweets) [inline synthetic_data/generator_builder.py:43]: added new to profile a dataset

File: synthetic_data/generator_builder.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        if not cls.is_valid_data(profile):
+            raise ValueError(
+                "Profile object is invalid. The supported profile "
+                f"types are: {list(cls.valid_data_types.keys())}."
+            )
+
+        generator = cls.valid_data_types[profile.__class__]
+        if data is not None:
+            profile = generator.post_profile_processing_w_data(data, profile)
```
Comment (JGSweets) [inline synthetic_data/generator_builder.py:64]: added new to conduct in profile postprocessing as needed, e.g. convert categorical columns to have a correlation associated with them if they are text based

File: synthetic_data/generators.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+class TabularGenerator(BaseGenerator):
+    """Class for generating synthetic tabular data."""
+
+    def __init__(self, profile, seed=None, noise_level: float = 0.0):
+        """Initialize tabular generator object."""
+        super().__init__(profile, seed)
+        self.noise_level = noise_level
+
+    @classmethod
```
Comment (JGSweets) [inline synthetic_data/generators.py:19]: added new

File: synthetic_data/generators.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
+"""Contains generators for tabular, graph, and unstructured data profiles."""
```
Comment (JGSweets) [inline synthetic_data/generators.py:1]: Mostly a copy from redesign api branch

File: tests/data/graph.csv [hunk 10]
```diff
@@ -0,0 +1,201 @@
+node_id_dst, node_id_src, continuous_weight, categorical_status
```
Comment (JGSweets) [inline tests/data/graph.csv:1]: copied from redesign branch

File: tests/data/iris.csv [hunk 11]
```diff
@@ -0,0 +1,151 @@
+sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target
```
Comment (JGSweets) [inline tests/data/iris.csv:1]: copy from redesign branch

File: tests/data/tabular.csv [hunk 12]
```diff
@@ -0,0 +1,18 @@
+datetime,host,src,proto,type,srcport,destport,srcip,locale,localeabbr,postalcode,latitude,longitude,owner,comment,int_col
```
Comment (JGSweets) [inline tests/data/tabular.csv:1]: copy from redesign branch

File: tests/test_generator_builder.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertEqual(synthetic_data.number_of_nodes(), 278)
+
+    def test_invalid_config(self):
+        with self.assertRaises(
+            ValueError,
+            msg="Warning: profile doesn't match user setting.",
+        ):
+            Generator(config=1)
+
+    def test_no_profile_or_data(self):
```
Comment (JGSweets) [inline tests/test_generator_builder.py:82]: updated for new kwarg

**Review by tazitoo — APPROVED** [review]
LGTM - thanks for the updated functionality

**Pull request merged**